sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,356 @@
+from sympy.physics.vector.frame import ReferenceFrame
+from sympy.physics.mechanics.body import Body
+from sympy.physics.vector import Vector, dynamicsymbols, cross
+from abc import ABC, abstractmethod
+
+__all__ = ['Joint', 'PinJoint']
+
+
+class Joint(ABC):
+    """Abstract Base class for all specific joints.
+
+    Explanation
+    ===========
+
+    A joint subtracts degrees of freedom from a body.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can be created by
+    inheriting Joint class and defining all abstract functions.
+
+    Parameters
+    ==========
+
+    name : string
+        The Joint's name which makes it unique.
+    parent : Body
+        The parent body of joint.
+    child : Body
+        The child body of joint.
+    coordinates: List of dynamicsymbols/dynamicsymbol, optional
+        Coordinates of joint.
+    speeds : List of dynamicsymbols/dynamicsymbol, optional
+        Speeds of joint.
+    parent_joint_pos : Vector, optional
+        Defines the joint's point where the parent will be connected to child.
+        Default value is masscenter of Parent Body.
+    child_joint_pos : Vector, optional
+        Defines the joint's point where the child will be connected to parent.
+        Default value is masscenter of Child Body.
+    parent_axis : Vector, optional
+        Axis of parent frame which would be aligned with child's
+        axis. Default is x axis in parent's frame.
+    child_axis : Vector, optional
+        Axis of child frame which would be aligned with parent's
+        axis. Default is x axis in child's frame.
+
+    """
+
+    def __init__(self, name, parent, child, coordinates=None, speeds=None, parent_joint_pos=None,
+        child_joint_pos=None, parent_axis=None, child_axis=None):
+
+        if not isinstance(name, str):
+            raise TypeError('Supply a valid name.')
+        self._name = name
+
+        if not isinstance(parent, Body):
+            raise TypeError('Parent must be an instance of Body.')
+        self._parent = parent
+
+        if not isinstance(child, Body):
+            raise TypeError('Child must be an instance of Body.')
+        self._child = child
+
+        self._coordinates = self._generate_coordinates(coordinates)
+        self._speeds = self._generate_speeds(speeds)
+        self._kdes = self._generate_kdes()
+
+        self._parent_axis = self._axis(parent, parent_axis)
+        self._child_axis = self._axis(child, child_axis)
+
+        self._parent_joint = self._locate_joint_pos(parent,parent_joint_pos)
+        self._child_joint = self._locate_joint_pos(child, child_joint_pos)
+
+        self._orient_frames()
+        self._set_angular_velocity()
+        self._set_linear_velocity()
+
+    def __str__(self):
+        return self._name
```

Comment:
This would be nice if it printed "Pin Joint connecting the parent body A to child body B" or something of that nature. Or "Pin Joint, parent: A, child: B"